src/cs/Ssh.Tcp/Services/PortForwardingService.cs

```diff
@@ -772,19 +772,25 @@ public PortForwardingService(SshSession session)
 	{
 		cancellation.ThrowIfCancellationRequested();
 
-		LocalPortForwarder? forwarder;
 		lock (DisposablesLock)
 		{
-			if (!this.localForwarders.TryGetValue(forwardedPort, out forwarder))
+			LocalPortForwarder? forwarder;
+			if (this.localForwarders.TryGetValue(forwardedPort, out forwarder))
 			{
-				return Task.FromResult(false);
+				localForwarders.Remove(forwardedPort);
+				forwarder.Dispose();
+				return Task.FromResult(true);
 			}
+		}
 
-			localForwarders.Remove(forwardedPort);
+		var port = new ForwardedPort(forwardedPort, forwardedPort, isRemote: true);
+		if (RemoteForwardedPorts.Contains(port))
+		{
+			RemoteForwardedPorts.RemovePort(port);
```

Comment:
This seems like it's open to a time-to-check-to-time-of-use race but it's not clear to me what the right solution is 🤔 

It looks like we don't hold a lock around `RemoteForwardedPorts` changes currently. We unconditionally call `ForwardedPortsCollection.RemovePort` and change it to not throw if the port isn't in it Thoughts?

Reply:
> unconditionally call `ForwardedPortsCollection.RemovePort` and change it to not throw if the port isn't in it

I think that would be correct. I see other calls to `RemovePort()` are done from within `Disposed` event handlers so if it does throw that could cause unclean shutdown or maybe even hangs. While that exception might indicate a bug (trying to remove a port that wasn't there) it seems more robust to not throw in that case anyway.